controllers/hpa_util.go

```diff
@@ -85,21 +145,86 @@ func (r *HvpaReconciler) UpdateHpaWithRetries(namespace, name string, applyUpdat
+		log.V(4).Info("Hpa %s precondition doesn't hold, skip updating it.", namespace+"/"+name)
+		retryErr = nil
+	}
+
+	return hpa, retryErr
+}
 
 func (r *HvpaReconciler) syncHpaSpec(hpaList []*autoscaling.HorizontalPodAutoscaler, hvpa *autoscalingv1alpha1.Hvpa) error {
 	for _, hpa := range hpaList {
 		if hpaChanged := hpaSpecNeedChange(hvpa, hpa); hpaChanged {
-			_, err := r.UpdateHpaWithRetries(hpa.Namespace, hpa.Name,
-				func(hpaToUpdate *autoscaling.HorizontalPodAutoscaler) {
-					hpaToUpdate.Spec.MinReplicas = hvpa.Spec.Hpa.Template.Spec.MinReplicas
-					hpaToUpdate.Spec.MaxReplicas = hvpa.Spec.Hpa.Template.Spec.MaxReplicas
-					hpaToUpdate.Spec.Metrics = hvpa.Spec.Hpa.Template.Spec.Metrics
-					hpaToUpdate.Spec.ScaleTargetRef = *hvpa.Spec.TargetRef
-				})
-			if err != nil {
-				return fmt.Errorf("error in updating hpaTemplateSpec to hpa %q: %v", hpa.Name, err)
+			if !r.IsAutoscalingV2Enabled {
+				_, err := r.UpdateHpaWithRetries(hpa.Namespace, hpa.Name,
+					func(hpaToUpdate *autoscaling.HorizontalPodAutoscaler) {
+						hpaToUpdate.Spec.MinReplicas = hvpa.Spec.Hpa.Template.Spec.MinReplicas
+						hpaToUpdate.Spec.MaxReplicas = hvpa.Spec.Hpa.Template.Spec.MaxReplicas
+						hpaToUpdate.Spec.Metrics = hvpa.Spec.Hpa.Template.Spec.Metrics
+						hpaToUpdate.Spec.ScaleTargetRef = *hvpa.Spec.TargetRef
+					})
+				if err != nil {
+					return fmt.Errorf("error in updating hpaTemplateSpec to hpa %q: %v", hpa.Name, err)
+				}
+				log.V(2).Info("HPA spec sync", "HPA", hpa.Name, "HVPA", hvpa.Namespace+"/"+hvpa.Name)
+			} else {
+				v2Hpa, err := r.Convert_v2beta1_HPA_to_v2(hpa)
+				if err != nil {
+					return err
+				}
+				v2HpaMetricSpecList := []autoscalingv2.MetricSpec{}
+				for _, metricSpec := range hvpa.Spec.Hpa.Template.Spec.Metrics {
+					v2HpaMetricSpec := &autoscalingv2.MetricSpec{}
+					internalHpaMetricSpec := &kubernetesinternalautoscaling.MetricSpec{}
+					err = r.Scheme.Convert(metricSpec, internalHpaMetricSpec, context.TODO())
+					if err != nil {
+						return err
+					}
+					err = r.Scheme.Convert(internalHpaMetricSpec, v2HpaMetricSpec, context.TODO())
+					if err != nil {
+						return err
+					}
+					v2HpaMetricSpecList = append(v2HpaMetricSpecList, *v2HpaMetricSpec)
+				}
+				v2CrossVersionObjectReference := &autoscalingv2.CrossVersionObjectReference{}
+				internalCrossVersionObjectReference := &kubernetesinternalautoscaling.CrossVersionObjectReference{}
+				err = r.Scheme.Convert(hvpa.Spec.TargetRef, internalCrossVersionObjectReference, context.TODO())
+				if err != nil {
+					return err
+				}
+				err = r.Scheme.Convert(internalCrossVersionObjectReference, v2CrossVersionObjectReference, context.TODO())
+				if err != nil {
+					return err
+				}
+				_, err = r.UpdateHpaV2WithRetries(v2Hpa.Namespace, v2Hpa.Name,
+					func(hpaToUpdate *autoscalingv2.HorizontalPodAutoscaler) {
+						hpaToUpdate.Spec.MinReplicas = hvpa.Spec.Hpa.Template.Spec.MinReplicas
```

Comment:
You're right. I do see all the assignments now.

I can't tell what happened, as the diff display has changed (likely because of the review comments we both added, or possibly because of the minor edits made based on the review). I guess it must have been a case of the diff incorrectly matching two similar but unrelated sections of code...